src/main/java/frc/robot/subsystems/swerve/module/SwerveModule.kt

```diff
@@ -88,10 +116,50 @@ class SwerveModule(
      * @param desiredState The desired state for the module.
      */
     fun applyState(desiredState: SwerveModuleState) {
+        this.desiredState = desiredState
+        
         desiredState.optimize(inputs.absoluteTurnPosition)
         desiredState.cosineScale(inputs.absoluteTurnPosition)
 
         io.setTurnPosition(desiredState.angle)
         io.setDriveVelocity(desiredState.speedMetersPerSecond measuredIn Units.MetersPerSecond)
     }
+
+    fun characterizeDriveVoltage(volts: Voltage) {
+        io.setTurnPosition(Rotation2d())
+        io.setDriveVoltage(volts)
+    }
+
+    fun characterizeSteerVoltage(volts: Voltage) {
+        io.setTurnVoltage(volts)
+        io.setDriveVoltage(Volts.zero())
+    }
+
+    fun characterizeCurrent(current: Current){
+        io.setTurnPosition(Rotation2d())
+        io.setDriveCurrent(current)
+    }
+
+    private companion object {
+        val driveTorqueGains: ControllerGains = ControllerGains(
+            PIDGains(
+                kP = 10.0
+            ),
+            FeedforwardGains(
+                kV = 1.15
```

Comment:
Did we verify this was right? I thought we said that kV should be much smaller than 1.15 for torque control